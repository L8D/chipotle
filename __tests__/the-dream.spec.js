'use strict';

/* globals describe, it, expect */

const Printer = require('graphql-tag/printer');
const gql = require('../lib');
const QueryShape = require('../lib/query-shape');

describe('the dream GQL interface', () => {
  const theId = 'foo';

  const photoIds = ['123', 'abc', 'xyz'];

  const pricingBreakdownFragment = gql.fragment`
    total
  `;

  const albumMembershipFragment = gql.fragment`
    album {
      name

      pricing_breakdown(photo_ids: ${gql.value(photoIds)}) {
        ...${pricingBreakdownFragment}
      }
    }
  `;

  const anotherAlbumMembershipFragment = gql.fragment`
    album {
      ... on IRecord {
        id
      }

      ... on Album {
        name
      }
    }

    account {
      id
      name
    }
  `;

  const queryFragment = gql.fragment`
    albumMembership(id: ${gql.value(theId)}) {
      ...${albumMembershipFragment}
      ...${anotherAlbumMembershipFragment}
    }
  `;

  const query = gql.query`
    query theQuery {
      ...${queryFragment}
    }
  `;

  const queryShape = QueryShape.fromDocumentAST(query);
  const flatAST = QueryShape.printQuerySet(queryShape.querySet);

  it('should match AST snapshot', () => {
    expect(query).toMatchSnapshot();
  });

  it('should flatten AST stuff', () => {
    expect(queryShape).toMatchSnapshot();
  });

  it('should make a new AST', () => {
    expect(flatAST).toMatchSnapshot();
  });

  it('should serialize a generated AST', () => {
    expect(Printer.print(flatAST)).toMatchSnapshot();
  });

  it('should merge fragments', () => {
    expect(gql.mergeFragments([
      albumMembershipFragment,
      anotherAlbumMembershipFragment
    ])).toMatchSnapshot();
  });
});
