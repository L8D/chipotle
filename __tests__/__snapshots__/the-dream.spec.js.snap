exports[`the dream GQL interface should flatten AST stuff 1`] = `
Object {
  "querySet": Object {
    "theQuery": Object {
      "fields": Object {
        "albumMembership": Object {
          "arguments": Array [
            Object {
              "kind": "Argument",
              "loc": null,
              "name": Object {
                "kind": "Name",
                "loc": null,
                "value": "id",
              },
              "value": Object {
                "kind": "StringValue",
                "loc": null,
                "value": "foo",
              },
            },
          ],
          "fields": Object {
            "account": Object {
              "arguments": Array [],
              "fields": Object {
                "id": Object {
                  "arguments": Array [],
                  "fields": null,
                  "name": "id",
                  "typeCondition": undefined,
                },
                "name": Object {
                  "arguments": Array [],
                  "fields": null,
                  "name": "name",
                  "typeCondition": undefined,
                },
              },
              "name": "account",
              "typeCondition": undefined,
            },
            "album": Object {
              "arguments": Array [],
              "fields": Object {
                "id": Object {
                  "arguments": Array [],
                  "fields": null,
                  "name": "id",
                  "typeCondition": Object {
                    "kind": "NamedType",
                    "loc": null,
                    "name": Object {
                      "kind": "Name",
                      "loc": null,
                      "value": "IRecord",
                    },
                  },
                },
                "name": Object {
                  "arguments": Array [],
                  "fields": null,
                  "name": "name",
                  "typeCondition": Object {
                    "kind": "NamedType",
                    "loc": null,
                    "name": Object {
                      "kind": "Name",
                      "loc": null,
                      "value": "Album",
                    },
                  },
                },
                "pricing_breakdown": Object {
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "loc": null,
                      "name": Object {
                        "kind": "Name",
                        "loc": null,
                        "value": "photo_ids",
                      },
                      "value": Object {
                        "kind": "ListValue",
                        "loc": null,
                        "values": Array [
                          Object {
                            "kind": "StringValue",
                            "loc": null,
                            "value": "123",
                          },
                          Object {
                            "kind": "StringValue",
                            "loc": null,
                            "value": "abc",
                          },
                          Object {
                            "kind": "StringValue",
                            "loc": null,
                            "value": "xyz",
                          },
                        ],
                      },
                    },
                  ],
                  "fields": Object {
                    "total": Object {
                      "arguments": Array [],
                      "fields": null,
                      "name": "total",
                      "typeCondition": undefined,
                    },
                  },
                  "name": "pricing_breakdown",
                  "typeCondition": undefined,
                },
              },
              "name": "album",
              "typeCondition": undefined,
            },
          },
          "name": "albumMembership",
          "typeCondition": undefined,
        },
      },
      "name": "theQuery",
    },
  },
}
`;

exports[`the dream GQL interface should make a new AST 1`] = `
Object {
  "definitions": Array [
    Object {
      "kind": "OperationDefinition",
      "name": Object {
        "kind": "Name",
        "value": "theQuery",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": null,
            "arguments": Array [
              Object {
                "kind": "Argument",
                "loc": null,
                "name": Object {
                  "kind": "Name",
                  "loc": null,
                  "value": "id",
                },
                "value": Object {
                  "kind": "StringValue",
                  "loc": null,
                  "value": "foo",
                },
              },
            ],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "albumMembership",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "album",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "kind": "InlineFragment",
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": null,
                            },
                          ],
                        },
                        "typeCondition": Object {
                          "kind": "NamedType",
                          "loc": null,
                          "name": Object {
                            "kind": "Name",
                            "loc": null,
                            "value": "Album",
                          },
                        },
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [
                          Object {
                            "kind": "Argument",
                            "loc": null,
                            "name": Object {
                              "kind": "Name",
                              "loc": null,
                              "value": "photo_ids",
                            },
                            "value": Object {
                              "kind": "ListValue",
                              "loc": null,
                              "values": Array [
                                Object {
                                  "kind": "StringValue",
                                  "loc": null,
                                  "value": "123",
                                },
                                Object {
                                  "kind": "StringValue",
                                  "loc": null,
                                  "value": "abc",
                                },
                                Object {
                                  "kind": "StringValue",
                                  "loc": null,
                                  "value": "xyz",
                                },
                              ],
                            },
                          },
                        ],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "pricing_breakdown",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "total",
                              },
                              "selectionSet": null,
                            },
                          ],
                        },
                      },
                      Object {
                        "kind": "InlineFragment",
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": null,
                            },
                          ],
                        },
                        "typeCondition": Object {
                          "kind": "NamedType",
                          "loc": null,
                          "name": Object {
                            "kind": "Name",
                            "loc": null,
                            "value": "IRecord",
                          },
                        },
                      },
                    ],
                  },
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "account",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "selectionSet": null,
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
    },
  ],
  "kind": "Document",
}
`;

exports[`the dream GQL interface should match AST snapshot 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "loc": null,
      "name": Object {
        "kind": "Name",
        "loc": null,
        "value": "theQuery",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "loc": null,
        "selections": Array [
          Object {
            "directives": Array [],
            "kind": "InlineFragment",
            "loc": null,
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": null,
              "selections": Array [
                Object {
                  "alias": null,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "loc": null,
                      "name": Object {
                        "kind": "Name",
                        "loc": null,
                        "value": "id",
                      },
                      "value": Object {
                        "kind": "StringValue",
                        "loc": null,
                        "value": "foo",
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": null,
                  "name": Object {
                    "kind": "Name",
                    "loc": null,
                    "value": "albumMembership",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "loc": null,
                    "selections": Array [
                      Object {
                        "directives": Array [],
                        "kind": "InlineFragment",
                        "loc": null,
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "loc": null,
                          "selections": Array [
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "loc": null,
                              "name": Object {
                                "kind": "Name",
                                "loc": null,
                                "value": "album",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "loc": null,
                                "selections": Array [
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "loc": null,
                                    "name": Object {
                                      "kind": "Name",
                                      "loc": null,
                                      "value": "name",
                                    },
                                    "selectionSet": null,
                                  },
                                  Object {
                                    "alias": null,
                                    "arguments": Array [
                                      Object {
                                        "kind": "Argument",
                                        "loc": null,
                                        "name": Object {
                                          "kind": "Name",
                                          "loc": null,
                                          "value": "photo_ids",
                                        },
                                        "value": Object {
                                          "kind": "ListValue",
                                          "loc": null,
                                          "values": Array [
                                            Object {
                                              "kind": "StringValue",
                                              "loc": null,
                                              "value": "123",
                                            },
                                            Object {
                                              "kind": "StringValue",
                                              "loc": null,
                                              "value": "abc",
                                            },
                                            Object {
                                              "kind": "StringValue",
                                              "loc": null,
                                              "value": "xyz",
                                            },
                                          ],
                                        },
                                      },
                                    ],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "loc": null,
                                    "name": Object {
                                      "kind": "Name",
                                      "loc": null,
                                      "value": "pricing_breakdown",
                                    },
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "loc": null,
                                      "selections": Array [
                                        Object {
                                          "directives": Array [],
                                          "kind": "InlineFragment",
                                          "loc": null,
                                          "selectionSet": Object {
                                            "kind": "SelectionSet",
                                            "loc": null,
                                            "selections": Array [
                                              Object {
                                                "alias": null,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "loc": null,
                                                "name": Object {
                                                  "kind": "Name",
                                                  "loc": null,
                                                  "value": "total",
                                                },
                                                "selectionSet": null,
                                              },
                                            ],
                                          },
                                          "typeCondition": null,
                                        },
                                      ],
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                        },
                        "typeCondition": null,
                      },
                      Object {
                        "directives": Array [],
                        "kind": "InlineFragment",
                        "loc": null,
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "loc": null,
                          "selections": Array [
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "loc": null,
                              "name": Object {
                                "kind": "Name",
                                "loc": null,
                                "value": "album",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "loc": null,
                                "selections": Array [
                                  Object {
                                    "directives": Array [],
                                    "kind": "InlineFragment",
                                    "loc": null,
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "loc": null,
                                      "selections": Array [
                                        Object {
                                          "alias": null,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "loc": null,
                                          "name": Object {
                                            "kind": "Name",
                                            "loc": null,
                                            "value": "id",
                                          },
                                          "selectionSet": null,
                                        },
                                      ],
                                    },
                                    "typeCondition": Object {
                                      "kind": "NamedType",
                                      "loc": null,
                                      "name": Object {
                                        "kind": "Name",
                                        "loc": null,
                                        "value": "IRecord",
                                      },
                                    },
                                  },
                                  Object {
                                    "directives": Array [],
                                    "kind": "InlineFragment",
                                    "loc": null,
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "loc": null,
                                      "selections": Array [
                                        Object {
                                          "alias": null,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "loc": null,
                                          "name": Object {
                                            "kind": "Name",
                                            "loc": null,
                                            "value": "name",
                                          },
                                          "selectionSet": null,
                                        },
                                      ],
                                    },
                                    "typeCondition": Object {
                                      "kind": "NamedType",
                                      "loc": null,
                                      "name": Object {
                                        "kind": "Name",
                                        "loc": null,
                                        "value": "Album",
                                      },
                                    },
                                  },
                                ],
                              },
                            },
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "loc": null,
                              "name": Object {
                                "kind": "Name",
                                "loc": null,
                                "value": "account",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "loc": null,
                                "selections": Array [
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "loc": null,
                                    "name": Object {
                                      "kind": "Name",
                                      "loc": null,
                                      "value": "id",
                                    },
                                    "selectionSet": null,
                                  },
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "loc": null,
                                    "name": Object {
                                      "kind": "Name",
                                      "loc": null,
                                      "value": "name",
                                    },
                                    "selectionSet": null,
                                  },
                                ],
                              },
                            },
                          ],
                        },
                        "typeCondition": null,
                      },
                    ],
                  },
                },
              ],
            },
            "typeCondition": null,
          },
        ],
      },
      "variableDefinitions": Array [],
    },
  ],
  "kind": "Document",
  "loc": null,
}
`;

exports[`the dream GQL interface should merge fragments 1`] = `
ChipotleFragment {
  "ast": Object {
    "definitions": Array [
      Object {
        "directives": Array [],
        "kind": "FragmentDefinition",
        "loc": null,
        "name": Object {
          "kind": "Name",
          "loc": null,
          "value": "temporaryFragment",
        },
        "selectionSet": Object {
          "kind": "SelectionSet",
          "loc": null,
          "selections": Array [
            Object {
              "directives": Array [],
              "kind": "InlineFragment",
              "loc": null,
              "selectionSet": Object {
                "kind": "SelectionSet",
                "loc": null,
                "selections": Array [
                  Object {
                    "alias": null,
                    "arguments": Array [],
                    "directives": Array [],
                    "kind": "Field",
                    "loc": null,
                    "name": Object {
                      "kind": "Name",
                      "loc": null,
                      "value": "album",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "loc": null,
                      "selections": Array [
                        Object {
                          "alias": null,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "loc": null,
                          "name": Object {
                            "kind": "Name",
                            "loc": null,
                            "value": "name",
                          },
                          "selectionSet": null,
                        },
                        Object {
                          "alias": null,
                          "arguments": Array [
                            Object {
                              "kind": "Argument",
                              "loc": null,
                              "name": Object {
                                "kind": "Name",
                                "loc": null,
                                "value": "photo_ids",
                              },
                              "value": Object {
                                "kind": "ListValue",
                                "loc": null,
                                "values": Array [
                                  Object {
                                    "kind": "StringValue",
                                    "loc": null,
                                    "value": "123",
                                  },
                                  Object {
                                    "kind": "StringValue",
                                    "loc": null,
                                    "value": "abc",
                                  },
                                  Object {
                                    "kind": "StringValue",
                                    "loc": null,
                                    "value": "xyz",
                                  },
                                ],
                              },
                            },
                          ],
                          "directives": Array [],
                          "kind": "Field",
                          "loc": null,
                          "name": Object {
                            "kind": "Name",
                            "loc": null,
                            "value": "pricing_breakdown",
                          },
                          "selectionSet": Object {
                            "kind": "SelectionSet",
                            "loc": null,
                            "selections": Array [
                              Object {
                                "directives": Array [],
                                "kind": "InlineFragment",
                                "loc": null,
                                "selectionSet": Object {
                                  "kind": "SelectionSet",
                                  "loc": null,
                                  "selections": Array [
                                    Object {
                                      "alias": null,
                                      "arguments": Array [],
                                      "directives": Array [],
                                      "kind": "Field",
                                      "loc": null,
                                      "name": Object {
                                        "kind": "Name",
                                        "loc": null,
                                        "value": "total",
                                      },
                                      "selectionSet": null,
                                    },
                                  ],
                                },
                                "typeCondition": null,
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
              "typeCondition": null,
            },
            Object {
              "directives": Array [],
              "kind": "InlineFragment",
              "loc": null,
              "selectionSet": Object {
                "kind": "SelectionSet",
                "loc": null,
                "selections": Array [
                  Object {
                    "alias": null,
                    "arguments": Array [],
                    "directives": Array [],
                    "kind": "Field",
                    "loc": null,
                    "name": Object {
                      "kind": "Name",
                      "loc": null,
                      "value": "album",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "loc": null,
                      "selections": Array [
                        Object {
                          "directives": Array [],
                          "kind": "InlineFragment",
                          "loc": null,
                          "selectionSet": Object {
                            "kind": "SelectionSet",
                            "loc": null,
                            "selections": Array [
                              Object {
                                "alias": null,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "loc": null,
                                "name": Object {
                                  "kind": "Name",
                                  "loc": null,
                                  "value": "id",
                                },
                                "selectionSet": null,
                              },
                            ],
                          },
                          "typeCondition": Object {
                            "kind": "NamedType",
                            "loc": null,
                            "name": Object {
                              "kind": "Name",
                              "loc": null,
                              "value": "IRecord",
                            },
                          },
                        },
                        Object {
                          "directives": Array [],
                          "kind": "InlineFragment",
                          "loc": null,
                          "selectionSet": Object {
                            "kind": "SelectionSet",
                            "loc": null,
                            "selections": Array [
                              Object {
                                "alias": null,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "loc": null,
                                "name": Object {
                                  "kind": "Name",
                                  "loc": null,
                                  "value": "name",
                                },
                                "selectionSet": null,
                              },
                            ],
                          },
                          "typeCondition": Object {
                            "kind": "NamedType",
                            "loc": null,
                            "name": Object {
                              "kind": "Name",
                              "loc": null,
                              "value": "Album",
                            },
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "alias": null,
                    "arguments": Array [],
                    "directives": Array [],
                    "kind": "Field",
                    "loc": null,
                    "name": Object {
                      "kind": "Name",
                      "loc": null,
                      "value": "account",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "loc": null,
                      "selections": Array [
                        Object {
                          "alias": null,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "loc": null,
                          "name": Object {
                            "kind": "Name",
                            "loc": null,
                            "value": "id",
                          },
                          "selectionSet": null,
                        },
                        Object {
                          "alias": null,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "loc": null,
                          "name": Object {
                            "kind": "Name",
                            "loc": null,
                            "value": "name",
                          },
                          "selectionSet": null,
                        },
                      ],
                    },
                  },
                ],
              },
              "typeCondition": null,
            },
          ],
        },
        "typeCondition": Object {
          "kind": "NamedType",
          "loc": null,
          "name": Object {
            "kind": "Name",
            "loc": null,
            "value": "Nothing",
          },
        },
      },
    ],
    "kind": "Document",
    "loc": null,
  },
  "fragments": undefined,
}
`;

exports[`the dream GQL interface should serialize a generated AST 1`] = `
"query theQuery {
  albumMembership(id: \"foo\") {
    album {
      ... on Album {
        name
      }
      pricing_breakdown(photo_ids: [\"123\", \"abc\", \"xyz\"]) {
        total
      }
      ... on IRecord {
        id
      }
    }
    account {
      id
      name
    }
  }
}
"
`;
